File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private boolean isValid = true;
    private boolean dfs(TreeNode root, int leftBound, int rightBound){
        if(root == null) {
            return true;
        }
       if(root.val <= leftBound || root.val >= rightBound) return false;
       return dfs(root.left, leftBound, root.val) && dfs(root.right, root.val, rightBound);
    }

    public boolean isValidBST(TreeNode root) {
        if(root == null) {
            return true;
        }
        return dfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    int res = int.MaxValue;
    int cnt = 0;
    int kth = 1;
    private void dfs(TreeNode root)
    {
        if(root == null) {
            return;
        }
        dfs(root.left);
        cnt++;
        if(cnt == kth && res == int.MaxValue) {
            res = root.val;
            return;
        }
        dfs(root.right);
    }
    public int KthSmallest(TreeNode root, int k) {
        kth = k;
        dfs(root);
        return res;
    }
}
